domain/port/service/watcher.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+
+	"github.com/juju/collections/transform"
+
+	"github.com/juju/juju/core/changestream"
+	"github.com/juju/juju/core/database"
+	"github.com/juju/juju/core/watcher"
+	"github.com/juju/juju/core/watcher/eventsource"
+)
+
+// WatcherFactory describes methods for creating watchers.
+type WatcherFactory interface {
+	// NewNamespaceMapperWatcher returns a new namespace watcher
+	// for events based on the input change mask and mapper.
+	NewNamespaceMapperWatcher(
+		namespace string,
+		changeMask changestream.ChangeType,
+		initialStateQuery eventsource.NamespaceQuery,
+		mapper eventsource.Mapper,
+	) (watcher.StringsWatcher, error)
+
+	// NewNamespaceNotifyMapperWatcher returns a new namespace notify watcher
+	// for events based on the input change mask and mapper.
+	NewNamespaceNotifyMapperWatcher(
+		namespace string, changeMask changestream.ChangeType, mapper eventsource.Mapper,
+	) (watcher.NotifyWatcher, error)
+}
+
+// WatcherState describes the methods that the service needs for it's watchers.
+type WatcherState interface {
+	// WatchOpenedPortsTable returns the name of the table that should be watched
+	WatchOpenedPortsTable() string
+
+	// InitialWatchOpenedPortsStatement returns the query to load the initial event
+	// for the WatchOpenedPorts watcher
+	InitialWatchOpenedPortsStatement() string
+
+	// GetMachinesForEndpoints returns map from endpoint uuids to the uuids of
+	// the machines which host that endpoint for each provided endpoint uuid.
+	GetMachinesForEndpoints(ctx context.Context, endpointUUIDs []string) (map[string]string, error)
+
+	// GetApplicationsForEndpoints returns map from endpoint uuids to the uuids
+	// of the applications which host that endpoint for each provided endpoint uuid.
```

Comment:
Agree! I actually added strong types for uuids in my other PR here:
https://github.com/juju/juju/pull/18147

However, since that one is currently on hold for a bit, I've elected not to try and port that bit over quite yet, I'm waiting to see what the conclusion there is.

Reply:
This is defo a change we need, especially with all the type-swapping